LLEXT: Add reloc support for `R_ARM_THM_CALL` (BL and BLX instructions)

bjarki-andreasen opened the pull request:
Add support for the relocation type R_ARM_ARM_THM_CALL which is
produced for the ARM Thumb BL and BLX (branch immediate)
instructions.

These instructions are used for non-static functions like
```
void test1(void)
{
}

void main(void)
{
        test1();
}
```
Without support for this relocation, test1() has to be static, otherwise the extension simply gets stuck in a forever loop.

The PR also adds a test which generate this relocation type. The output from the test is as follows:
```
enter
relative jump 1
relative jump 2
relative jump 3
relative jump 4
relative jump 5
exit
```
The functions are placed in an order which hopefully makes the compiler place them in an order resulting in relative jumps in both directions, but we don't actually have control over that :)

bjarki-andreasen committed:
llext: arm: Add R_ARM_ARM_THM_CALL reloc support

Add support for the relocation type R_ARM_ARM_THM_CALL which is
produced for the ARM Thumb BL and BLX (branch immediate)
instructions.

These instructions are used for non-static functions like

void test1(void)
{
}

void main(void)
{
        test1();
}

Without support for this relocation, test1() has to be static.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
tests: llext: add relative jump test ext.c

Validate the new relocations for BL and BLX instructions by
creating a new test extension which contains a chain of global
functions in a pseudo random order to (hopefully) generate
relative jumps in both positive and negative directions.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

teburd commented:
Related PR #68787

teburd reviewed (commented):
Please make a new test extension for this rather than adding to the logging extension, otherwise it looks great

pillo79 commented on include/zephyr/llext/elf.h:
```diff
@@ -371,6 +371,7 @@ struct elf64_rela {
 #define R_ARM_ABS32 2
 #define R_ARM_REL32 3
 #define R_ARM_COPY 4
+#define R_ARM_ARM_THM_CALL 10
```
Too many arms I think! :wink:

bjarki-trackunit replied:
I do agree :)

pillo79 commented on arch/arm/core/elf.c:
```diff
@@ -29,6 +68,22 @@ void arch_elf_relocate(elf_rela_t *rel, uintptr_t opaddr, uintptr_t opval)
 		*((uint32_t *)opaddr) = (uint32_t)opval;
 		break;
+	case R_ARM_ARM_THM_CALL:
+		/*
+		 * The branch offset (addend) is encoded into the value at opaddr.
+		 * Start by subtracting the raw value of opaddr added to opval
+		 */
+		opval -= *((uint32_t *)opaddr);
```
This is definitely highlighting an issue with the current code, and I think it's time to fix it. 
The source is [this calculation](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/llext/llext.c#L771-L774), which assumes `*op_loc` contains the offset; that is in fact true only for `R_ARM_ABS32` relocs.
Can you fix this in a commit that moves this addend from `llext.c` to the `R_ARM_ABS32` case above? That way you will not need this odd bit of magic.

bjarki-trackunit replied:
Absolutely :) I was going to create a PR specifically for this, may as well fix it now

teburd replied:
Thank you

pillo79 reviewed (commented):
Thanks, very useful! I would like to ask for two quick fixes.

bjarki-andreasen committed:
llext: arch_elf_relocate: pass opval unmodified

The opval argument of arch_elf_relocate() was modified by
adding the value stored at opaddr before passing it to
arch_elf_relocate(). This presumed that the addend would
always be stored as a raw value at opaddr, which is not the
case for all relocation types.

This PR modifies opval to be the absolute address of opval,
and moves the addition of the addend from llext_link_plt()
to the implementation of arch_elf_relocate().

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

laurenmurphyx64 reviewed (dismissed):
Tested it out on my FRDM-K64F

bjarki-andreasen committed:
tests: llext: Update testcase to filter on ARCH_HAS_USERSPACE

Change the filter from CPU_HAS_MPU to ARCH_HAS_USERSPACE to
when filtering boards which support userspace.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-trackunit commented:
I'm adding an additional commit to update the `testcase.yaml` to better filter tests requiring USERSPACE for boards which support it. This is to avoid this CI build error:
```
warning: USERSPACE (defined at arch/Kconfig:274) was assigned the value 'y' but got the value 'n'.
Check these unsatisfied dependencies: ARCH_HAS_USERSPACE (=n), SRAM_REGION_PERMISSIONS (=n)
```
when building for `lpcxpresso55s69_lpc55s69_cpu0_ns`

I also rebased to hopefully get rid of a CI build error for mcumgr :)

pillo79 reviewed (approved):
Thanks for your help!

OT: Love how easy it is to add a test after #69689 got merged :+1:
